File: tests/game_objects/systems/test_attacks.py
"""Tests all classes and functions in game_objects/systems/attacks.py."""
from __future__ import annotations

# Builtin
from typing import TYPE_CHECKING

# Pip
import pytest

# Custom
from hades.game_objects.base import AttackAlgorithms
from hades.game_objects.components import Armour, Attacks, Health
from hades.game_objects.registry import Registry, RegistryError
from hades.game_objects.steering import Vec2d
from hades.game_objects.systems.attacks import AttackSystem
from hades.game_objects.systems.attributes import GameObjectAttributeSystem

if TYPE_CHECKING:
    from collections.abc import Callable

__all__ = ()


@pytest.fixture()
def registry() -> Registry:
    """Create a registry for use in testing.

    Returns:
        The registry for use in testing.
    """
    return Registry()


@pytest.fixture()
def attack_system_factory(
    registry: Registry,
) -> Callable[[list[AttackAlgorithms]], AttackSystem]:
    """Create an attack system factory for use in testing.

    Args:
        registry: The registry for use in testing.

    Returns:
        The attack system factory for use in testing.
    """

    def wrap(enabled_attacks: list[AttackAlgorithms]) -> AttackSystem:
        attack_system = AttackSystem(registry)
        game_object_id = registry.create_game_object(
            Attacks(enabled_attacks),
            kinematic=True,
        )
        registry.add_system(attack_system)
        registry.add_system(GameObjectAttributeSystem(registry))
        registry.get_kinematic_object_for_game_object(game_object_id).rotation = 180
        return attack_system

    return wrap


@pytest.fixture()
def targets(registry: Registry) -> list[int]:
    """Create a list of targets for use in testing.

    Args:
        registry: The registry for use in testing.

    Returns:
        The list of targets for use in testing.
    """

    def create_target(position: Vec2d) -> int:
        target = registry.create_game_object(
            Health(50, -1),
            Armour(0, -1),
            kinematic=True,
        )
        registry.get_kinematic_object_for_game_object(target).position = position
        return target

    return [
        create_target(Vec2d(-20, -100)),
        create_target(Vec2d(20, 60)),
        create_target(Vec2d(-200, 100)),
        create_target(Vec2d(100, -100)),
        create_target(Vec2d(-100, -99)),
        create_target(Vec2d(0, -200)),
        create_target(Vec2d(0, -192)),
        create_target(Vec2d(0, 0)),
    ]


def test_attacks_init(
    attack_system_factory: Callable[[list[AttackAlgorithms]], AttackSystem],
) -> None:
    """Test that the attacks component is initialised correctly.

    Args:
        attack_system_factory: The attack system factory for use in testing.
    """
    assert (
        repr(attack_system_factory([AttackAlgorithms.AREA_OF_EFFECT_ATTACK]))
        == "<AttackSystem (Description=`Provides facilities to manipulate attack"
        " components.`)>"
    )


def test_attacks_do_attack_area_of_effect_attack(
    attack_system_factory: Callable[[list[AttackAlgorithms]], AttackSystem],
    targets: list[int],
) -> None:
    """Test that performing an area of effect attack works correctly.

    Args:
        attack_system_factory: The attack system factory for use in testing.
        targets: The list of targets for use in testing.
    """
    attack_system = attack_system_factory([AttackAlgorithms.AREA_OF_EFFECT_ATTACK])
    assert attack_system.do_attack(8, targets) == {}
    assert (
        attack_system.registry.get_component_for_game_object(targets[0], Health).value
        == 40
    )
    assert (
        attack_system.registry.get_component_for_game_object(targets[1], Health).value
        == 40
    )
    assert (
        attack_system.registry.get_component_for_game_object(targets[2], Health).value
        == 50
    )
    assert (
        attack_system.registry.get_component_for_game_object(targets[3], Health).value
        == 40
    )
    assert (
        attack_system.registry.get_component_for_game_object(targets[4], Health).value
        == 40
    )
    assert (
        attack_system.registry.get_component_for_game_object(targets[5], Health).value
        == 50
    )
    assert (
        attack_system.registry.get_component_for_game_object(targets[6], Health).value
        == 40
    )
    assert (
        attack_system.registry.get_component_for_game_object(targets[7], Health).value
        == 40
    )


def test_attacks_do_attack_melee_attack(
    attack_system_factory: Callable[[list[AttackAlgorithms]], AttackSystem],
    targets: list[int],
) -> None:
    """Test that performing a melee attack works correctly.

    Args:
        attack_system_factory: The attack system factory for use in testing.
        targets: The list of targets for use in testing.
    """
    attack_system = attack_system_factory([AttackAlgorithms.MELEE_ATTACK])
    assert attack_system.do_attack(8, targets) == {}
    assert (
        attack_system.registry.get_component_for_game_object(targets[0], Health).value
        == 40
    )
    assert (
        attack_system.registry.get_component_for_game_object(targets[1], Health).value
        == 50
    )
    assert (
        attack_system.registry.get_component_for_game_object(targets[2], Health).value
        == 50
    )
    assert (
        attack_system.registry.get_component_for_game_object(targets[3], Health).value
        == 40
    )
    assert (
        attack_system.registry.get_component_for_game_object(targets[4], Health).value
        == 50
    )
    assert (
        attack_system.registry.get_component_for_game_object(targets[5], Health).value
        == 50
    )
    assert (
        attack_system.registry.get_component_for_game_object(targets[6], Health).value
        == 40
    )
    assert (
        attack_system.registry.get_component_for_game_object(targets[7], Health).value
        == 40
    )


def test_attacks_do_attack_ranged_attack(
    attack_system_factory: Callable[[list[AttackAlgorithms]], AttackSystem],
) -> None:
    """Test that performing a ranged attack works correctly.

    Args:
        attack_system_factory: The attack system factory for use in testing.
    """
    attack_system = attack_system_factory([AttackAlgorithms.RANGED_ATTACK])
    assert attack_system.do_attack(0, []) == {
        "ranged_attack": (
            Vec2d(0, 0),
            -300.0,
            pytest.approx(0),  # This is due to floating point errors
        ),
    }


def test_attacks_do_attack_invalid_game_object_id(
    attack_system_factory: Callable[[list[AttackAlgorithms]], AttackSystem],
) -> None:
    """Test that an exception is raised if an invalid game object ID is provided.

    Args:
        attack_system_factory: The attack system factory for use in testing.
    """
    with pytest.raises(
        expected_exception=RegistryError,
        match="The game object ID `-1` is not registered with the registry.",
    ):
        attack_system_factory([]).do_attack(-1, [])


def test_attacks_previous_next_attack_single(
    attack_system_factory: Callable[[list[AttackAlgorithms]], AttackSystem],
) -> None:
    """Test that switching between attacks once works correctly.

    Args:
        attack_system_factory: The attack system factory for use in testing.
    """
    attack_system = attack_system_factory(
        [
            AttackAlgorithms.AREA_OF_EFFECT_ATTACK,
            AttackAlgorithms.MELEE_ATTACK,
            AttackAlgorithms.RANGED_ATTACK,
        ],
    )
    assert (
        attack_system.registry.get_component_for_game_object(0, Attacks).attack_state
        == 0
    )
    attack_system.next_attack(0)
    assert (
        attack_system.registry.get_component_for_game_object(0, Attacks).attack_state
        == 1
    )
    attack_system.previous_attack(0)
    assert (
        attack_system.registry.get_component_for_game_object(0, Attacks).attack_state
        == 0
    )


def test_attacks_previous_attack_multiple(
    attack_system_factory: Callable[[list[AttackAlgorithms]], AttackSystem],
) -> None:
    """Test that switching between attacks multiple times works correctly.

    Args:
        attack_system_factory: The attack system factory for use in testing.
    """
    attack_system = attack_system_factory(
        [
            AttackAlgorithms.AREA_OF_EFFECT_ATTACK,
            AttackAlgorithms.MELEE_ATTACK,
            AttackAlgorithms.RANGED_ATTACK,
        ],
    )
    assert (
        attack_system.registry.get_component_for_game_object(0, Attacks).attack_state
        == 0
    )
    attack_system.next_attack(0)
    assert (
        attack_system.registry.get_component_for_game_object(0, Attacks).attack_state
        == 1
    )
    attack_system.next_attack(0)
    assert (
        attack_system.registry.get_component_for_game_object(0, Attacks).attack_state
        == 2
    )
    attack_system.next_attack(0)
    assert (
        attack_system.registry.get_component_for_game_object(0, Attacks).attack_state
        == 2
    )
    attack_system.previous_attack(0)
    assert (
        attack_system.registry.get_component_for_game_object(0, Attacks).attack_state
        == 1
    )
    attack_system.previous_attack(0)
    assert (
        attack_system.registry.get_component_for_game_object(0, Attacks).attack_state
        == 0
    )
    attack_system.previous_attack(0)
    assert (
        attack_system.registry.get_component_for_game_object(0, Attacks).attack_state
        == 0
    )


def test_attacks_previous_next_attack_empty_attacks(
    attack_system_factory: Callable[[list[AttackAlgorithms]], AttackSystem],
) -> None:
    """Test that changing the attack state works correctly when there are no attacks.

    Args:
        attack_system_factory: The attack system factory for use in testing.
    """
    attack_system = attack_system_factory([])
    assert (
        attack_system.registry.get_component_for_game_object(0, Attacks).attack_state
        == 0
    )
    attack_system.next_attack(0)
    assert (
        attack_system.registry.get_component_for_game_object(0, Attacks).attack_state
        == -1
    )
    attack_system.previous_attack(0)
    assert (
        attack_system.registry.get_component_for_game_object(0, Attacks).attack_state
        == 0
    )


def test_attacks_previous_next_attack_invalid_game_object_id(
    attack_system_factory: Callable[[list[AttackAlgorithms]], AttackSystem],
) -> None:
    """Test that an exception is raised if an invalid game object ID is provided.

    Args:
        attack_system_factory: The attack system factory for use in testing.
    """
    attack_system = attack_system_factory([])
    with pytest.raises(
        expected_exception=RegistryError,
        match="The game object ID `-1` is not registered with the registry.",
    ):
        attack_system.previous_attack(-1)
    with pytest.raises(
        expected_exception=RegistryError,
        match="The game object ID `-1` is not registered with the registry.",
    ):
        attack_system.next_attack(-1)

File: hades_extensions/test/game_objects/systems/test_attacks.cpp
// External includes
#include "gtest/gtest.h"

// Custom includes
#include "macros.hpp"
#include "game_objects/stats.hpp"
#include "game_objects/systems/attacks.hpp"

// ----- FIXTURES ------------------------------
/// Implements the fixture for the AttackSystem tests.
class AttackSystemFixture : public testing::Test {
 protected:
  /// The registry that manages the game objects, components, and systems.
  Registry registry{};

  /// A list of targets for use in testing.
  std::vector<int> targets{};

  /// Set up the fixture for the tests.
  void SetUp() override {
    // The lambda function for creating a target
    auto create_target = [&](Vec2d position) {
      std::vector<std::unique_ptr<ComponentBase>> components;
      components.push_back(std::make_unique<Health>(50, -1));
      components.push_back(std::make_unique<Armour>(0, -1));
      int target = registry.create_game_object(true, std::move(components));
      registry.get_kinematic_object(target)->position = position;
      return target;
    };

    // Create the targets and add the attacks system
    targets = {
        create_target({-20, -100}),
        create_target({20, 60}),
        create_target({-200, 100}),
        create_target({100, -100}),
        create_target({-100, -99}),
        create_target({0, -200}),
        create_target({0, -192}),
        create_target({0, 0}),
    };
    registry.add_system(std::make_shared<AttackSystem>(registry));
    registry.add_system(std::make_shared<DamageSystem>(registry));
  }

  /// Create an attacks component for a game object.
  ///
  /// @tparam T - The type of the component or system.
  /// @param list - The initializer list to pass to the constructor.
  void create_attack_component(const std::vector<AttackAlgorithms> &enabled_attacks) {
    std::vector<std::unique_ptr<ComponentBase>> components;
    components.push_back(std::make_unique<Attacks>(enabled_attacks));
    int game_object_id = registry.create_game_object(true, std::move(components));
    registry.get_kinematic_object(game_object_id)->rotation = 180;
  }

  /// Get the attacks system from the registry.
  ///
  /// @return The attacks system.
  std::shared_ptr<AttackSystem> get_attacks_system() {
    return registry.find_system<AttackSystem>();
  }
};

/// Implements the fixture for the DamageSystem tests.
class DamageSystemFixture : public testing::Test {
 protected:
  /// The registry that manages the game objects, components, and systems.
  Registry registry{};

  /// Set up the fixture for the tests.
  void SetUp() override {
    registry.add_system(std::make_shared<DamageSystem>(registry));
  }

  /// Create health and armour attributes for use in testing.
  void create_health_and_armour_attributes() {
    std::vector<std::unique_ptr<ComponentBase>> components;
    components.push_back(std::make_unique<Health>(300, -1));
    components.push_back(std::make_unique<Armour>(100, -1));
    registry.create_game_object(false, std::move(components));
  }

  /// Get the damage system from the registry.
  ///
  /// @return The damage system.
  std::shared_ptr<DamageSystem> get_damage_system() {
    return registry.find_system<DamageSystem>();
  }
};

// ----- TESTS ----------------------------------
/// Test that performing an area of effect attack works correctly.
TEST_F(AttackSystemFixture, TestAttacksDoAreaOfEffectAttack) {
  create_attack_component({AttackAlgorithms::AreaOfEffect});
  get_attacks_system()->do_attack(8, targets);
  ASSERT_EQ(registry.get_component<Health>(targets[0])->get_value(), 40);
  ASSERT_EQ(registry.get_component<Health>(targets[1])->get_value(), 40);
  ASSERT_EQ(registry.get_component<Health>(targets[2])->get_value(), 50);
  ASSERT_EQ(registry.get_component<Health>(targets[3])->get_value(), 40);
  ASSERT_EQ(registry.get_component<Health>(targets[4])->get_value(), 40);
  ASSERT_EQ(registry.get_component<Health>(targets[5])->get_value(), 50);
  ASSERT_EQ(registry.get_component<Health>(targets[6])->get_value(), 40);
  ASSERT_EQ(registry.get_component<Health>(targets[7])->get_value(), 40);
}

/// Test that performing a melee attack works correctly.
TEST_F(AttackSystemFixture, TestAttacksDoMeleeAttack) {
  create_attack_component({AttackAlgorithms::Melee});
  get_attacks_system()->do_attack(8, targets);
  ASSERT_EQ(registry.get_component<Health>(targets[0])->get_value(), 40);
  ASSERT_EQ(registry.get_component<Health>(targets[1])->get_value(), 50);
  ASSERT_EQ(registry.get_component<Health>(targets[2])->get_value(), 50);
  ASSERT_EQ(registry.get_component<Health>(targets[3])->get_value(), 40);
  ASSERT_EQ(registry.get_component<Health>(targets[4])->get_value(), 50);
  ASSERT_EQ(registry.get_component<Health>(targets[5])->get_value(), 50);
  ASSERT_EQ(registry.get_component<Health>(targets[6])->get_value(), 40);
  ASSERT_EQ(registry.get_component<Health>(targets[7])->get_value(), 40);
}

/// Test that performing a ranged attack works correctly.
TEST_F(AttackSystemFixture, TestAttacksDoRangedAttack) {
  // This is due to floating point precision
  create_attack_component({AttackAlgorithms::Ranged});
  std::tuple<Vec2d, double, double> attack_result = get_attacks_system()->do_attack(8, targets).ranged_attack.value();
  ASSERT_EQ(get<0>(attack_result), Vec2d(0, 0));
  ASSERT_EQ(get<1>(attack_result), -300);
  ASSERT_NEAR(get<2>(attack_result), 0, 1e-13);
}

/// Test that an exception is raised if an invalid game object ID is provided.
TEST_F(AttackSystemFixture, TestAttacksDoAttackInvalidGameObjectId) {
  create_attack_component({AttackAlgorithms::Ranged});
  ASSERT_THROW_MESSAGE(get_attacks_system()->do_attack(-1, targets),
                       RegistryException,
                       "The game object `-1` is not registered with the registry.")
}

/// Test that switching between attacks once works correctly.
TEST_F(AttackSystemFixture, TestAttacksPreviousNextAttackSingle) {
  create_attack_component({AttackAlgorithms::AreaOfEffect, AttackAlgorithms::Melee, AttackAlgorithms::Ranged});
  ASSERT_EQ(registry.get_component<Attacks>(8)->attack_state, 0);
  get_attacks_system()->next_attack(8);
  ASSERT_EQ(registry.get_component<Attacks>(8)->attack_state, 1);
  get_attacks_system()->previous_attack(8);
  ASSERT_EQ(registry.get_component<Attacks>(8)->attack_state, 0);
}

/// Test that switching between attacks multiple times works correctly.
TEST_F(AttackSystemFixture, TestAttacksPreviousAttackMultiple) {
  create_attack_component({AttackAlgorithms::AreaOfEffect, AttackAlgorithms::Melee, AttackAlgorithms::Ranged});
  ASSERT_EQ(registry.get_component<Attacks>(8)->attack_state, 0);
  get_attacks_system()->next_attack(8);
  ASSERT_EQ(registry.get_component<Attacks>(8)->attack_state, 1);
  get_attacks_system()->next_attack(8);
  ASSERT_EQ(registry.get_component<Attacks>(8)->attack_state, 2);
  get_attacks_system()->next_attack(8);
  ASSERT_EQ(registry.get_component<Attacks>(8)->attack_state, 2);
  get_attacks_system()->previous_attack(8);
  ASSERT_EQ(registry.get_component<Attacks>(8)->attack_state, 1);
  get_attacks_system()->previous_attack(8);
  ASSERT_EQ(registry.get_component<Attacks>(8)->attack_state, 0);
  get_attacks_system()->previous_attack(8);
  ASSERT_EQ(registry.get_component<Attacks>(8)->attack_state, 0);
}

/// Test that changing the attack state works correctly when there are no attacks.
TEST_F(AttackSystemFixture, TestAttacksPreviousNextAttackEmptyAttacks) {
  create_attack_component({});
  ASSERT_EQ(registry.get_component<Attacks>(8)->attack_state, 0);
  get_attacks_system()->next_attack(8);
  ASSERT_EQ(registry.get_component<Attacks>(8)->attack_state, 0);
  get_attacks_system()->previous_attack(8);
  ASSERT_EQ(registry.get_component<Attacks>(8)->attack_state, 0);
}

/// Test that an exception is raised if an invalid game object ID is provided.
TEST_F(AttackSystemFixture, TestAttacksPreviousNextAttackInvalidGameObjectId) {
  create_attack_component({});
  ASSERT_THROW_MESSAGE(get_attacks_system()->next_attack(-1),
                       RegistryException,
                       "The game object `-1` is not registered with the registry.")
  ASSERT_THROW_MESSAGE(get_attacks_system()->previous_attack(-1),
                       RegistryException,
                       "The game object `-1` is not registered with the registry.")
}

/// Test that damage is dealt when health and armour are lower than damage.
TEST_F(DamageSystemFixture, TestGameObjectAttributeSystemDealDamageLowHealthArmour) {
  create_health_and_armour_attributes();
  get_damage_system()->deal_damage(0, 350);
  ASSERT_EQ(registry.get_component<Health>(0)->get_value(), 50);
  ASSERT_EQ(registry.get_component<Armour>(0)->get_value(), 0);
}

/// Test that no damage is dealt when armour is larger than damage.
TEST_F(DamageSystemFixture, TestGameObjectAttributeSystemDealDamageLargeArmour) {
  create_health_and_armour_attributes();
  get_damage_system()->deal_damage(0, 50);
  ASSERT_EQ(registry.get_component<Health>(0)->get_value(), 300);
  ASSERT_EQ(registry.get_component<Armour>(0)->get_value(), 50);
}

/// Test that no damage is dealt when damage is zero.
TEST_F(DamageSystemFixture, TestGameObjectAttributeSystemDealDamageZeroDamage) {
  create_health_and_armour_attributes();
  get_damage_system()->deal_damage(0, 0);
  ASSERT_EQ(registry.get_component<Health>(0)->get_value(), 300);
  ASSERT_EQ(registry.get_component<Armour>(0)->get_value(), 100);
}

/// Test that damage is dealt when armour is zero.
TEST_F(DamageSystemFixture, TestGameObjectAttributeSystemDealDamageZeroArmour) {
  create_health_and_armour_attributes();
  auto armour = registry.get_component<Armour>(0);
  armour->set_value(0);
  get_damage_system()->deal_damage(0, 100);
  ASSERT_EQ(registry.get_component<Health>(0)->get_value(), 200);
  ASSERT_EQ(armour->get_value(), 0);
}

/// Test that damage is dealt when health is zero.
TEST_F(DamageSystemFixture, TestGameObjectAttributeSystemDealDamageZeroHealth) {
  create_health_and_armour_attributes();
  auto health = registry.get_component<Health>(0);
  health->set_value(0);
  get_damage_system()->deal_damage(0, 50);
  ASSERT_EQ(health->get_value(), 0);
  ASSERT_EQ(registry.get_component<Armour>(0)->get_value(), 50);
}

/// Test that no damage is dealt when the attributes are not initialised.
TEST_F(DamageSystemFixture, TestGameObjectAttributeSystemDealDamageNonexistentAttributes) {
  registry.create_game_object(false, {});
  ASSERT_THROW_MESSAGE(get_damage_system()->deal_damage(0, 100),
                       RegistryException,
                       "The game object `0` is not registered with the registry.")
}

/// Test that an exception is raised if an invalid game object ID is provided.
TEST_F(DamageSystemFixture, TestGameObjectAttributeSystemDealDamageInvalidGameObjectId) {
  ASSERT_THROW_MESSAGE(get_damage_system()->deal_damage(-1, 100),
                       RegistryException,
                       "The game object `-1` is not registered with the registry.")
}
